pkg/covermerger/bq_csv_reader.go

```diff
@@ -77,48 +62,82 @@ func (r *bqCSVReader) InitNsRecords(ctx context.Context, ns, filePath, commit st
 	`, gsURI, ns, from.String(), to.String(), filePath, selectCommit))
 	job, err := q.Run(ctx)
 	if err != nil {
-		return fmt.Errorf("err during bigquery.Run: %w", err)
+		return nil, fmt.Errorf("err during bigquery.Run: %w", err)
 	}
 	status, err := job.Wait(ctx)
 	if err != nil {
-		return fmt.Errorf("err waiting for the bigquery.Job: %w", err)
+		return nil, fmt.Errorf("err waiting for the bigquery.Job: %w", err)
 	}
 	if status.Err() != nil {
-		return fmt.Errorf("bigquery job failed with status %w", status.Err())
+		return nil, fmt.Errorf("bigquery job failed with status %w", status.Err())
 	}
-	return r.initGCSFileReaders(ctx, gsBucket, gsPath)
+	return initGCSMultiReader(ctx, gsBucket, gsPath)
 }
 
-func (r *bqCSVReader) initGCSFileReaders(ctx context.Context, bucket, path string) error {
+func initGCSMultiReader(ctx context.Context, bucket, path string) (io.ReadCloser, error) {
 	var gcsClient gcs.Client
 	var err error
 	if gcsClient, err = gcs.NewClient(ctx); err != nil {
-		return fmt.Errorf("err creating gcs client: %w", err)
+		return nil, fmt.Errorf("err creating gcs client: %w", err)
 	}
 	var gcsFiles []*gcs.Object
 	if gcsFiles, err = gcsClient.ListObjects(bucket + "/" + path); err != nil {
-		return fmt.Errorf("err enumerating gcs files: %w", err)
+		return nil, fmt.Errorf("err enumerating gcs files: %w", err)
 	}
+	paths := []string{}
 	for _, obj := range gcsFiles {
-		var readCloser io.ReadCloser
-		if readCloser, err = gcsClient.FileReader(bucket + "/" + obj.Path); err != nil {
-			return fmt.Errorf("failed to get %s reader: %w", obj.Path, err)
-		}
-		r.closers = append(r.closers, readCloser)
-		r.gcsFiles = append(r.gcsFiles, readCloser)
+		paths = append(paths, bucket+"/"+obj.Path)
 	}
-	return nil
+	return &gcsGZIPMultiReader{
+		gcsClient: gcsClient,
+		gcsFiles:  paths,
+	}, nil
 }
 
-func (r *bqCSVReader) Reader() (io.Reader, error) {
-	var readers []io.Reader
-	for _, file := range r.gcsFiles {
-		gzipReaderCloser, err := gzip.NewReader(file)
-		if err != nil {
-			return nil, fmt.Errorf("err calling gzip.NewReader: %w", err)
+type gcsGZIPMultiReader struct {
+	gcsClient gcs.Client
+	gcsFiles  []string
+
+	curCloser io.Closer
+	curReader io.Reader
+}
+
+func (mr *gcsGZIPMultiReader) Read(p []byte) (n int, err error) {
```

Comment:
It would be nice to see some tests.

Another option (with a simpler implementation?) is to keep on using `io.MultiReader`, but add some lazy `io.Reader` implementation that would open the file on the first `Read()`.

Reply:
Lazy reader w/o lazy closer will help to solve the Attrs()==INTERNAL_ERROR problem.
Alone it will not help to solve the memory overhead problem.
Laze closer (wait for the first io.EOF and close) will help to solve the memory overuse problem.

But together laze reader+closer complexity looks higher than this function.
Logically this function is a modified copy-paste of the standard io.MultiReader.

One more alternative is to add one more dependency to get MultiReaderCloser. I decided to add 1 function instead.

Let's add function specific tests. It should be covered indirectly, but I also feel it is not enough.